### PR TITLE
Add CRUD for category

### PR DESCRIPTION
**Pull request by knurff**

### CODE REVIEW

File: src/main/java/com/example/springlabs/controllers/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+  @GetMapping("/**")
+  public Collection<CategoryDto> getSubcategories(HttpServletRequest request) {
+    List<String> categoriesNames = getCategoriesNames(request);
+    return categoryDtoMapper.createDtosSet(categoryService.getSubcategories(categoriesNames));
+  }
+
+  @PostMapping
+  @ResponseStatus(HttpStatus.CREATED)
+  public CategoryDto addCategory(@RequestBody CategoryDto categoryDto) {
```
Comment (xd6d) [inline src/main/java/com/example/springlabs/controllers/CategoryController.java:51]: Mb add smth like CategoryCreateDto without products and subCategories fields. I think new category immediatly after its creation does not have subCategories and products

File: src/main/java/com/example/springlabs/controllers/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        categoryService.updateCategory(categoriesNames, Long.parseLong(id),
+            categoryDtoMapper.createCategoryFromDto(categoryDto)));
+  }
+
+  @DeleteMapping("/**")
+  @ResponseStatus(HttpStatus.NO_CONTENT)
+  public void deleteCategory(HttpServletRequest request) {
+    List<String> categoriesNames = new ArrayList<>(getCategoriesNames(request));
+    String id = categoriesNames.remove(categoriesNames.size() - 1);
+        categoryService.deleteCategory(categoriesNames, Long.parseLong(id));
```
Comment (xd6d) [inline src/main/java/com/example/springlabs/controllers/CategoryController.java:82]: Same as update

**Review by xd6d — CHANGES_REQUESTED** [review]
Also, I would move ApplicationExceptionHandler to controllers package and ExceptionResponse to dtos package, or rename package exceptions to exception([link ](https://softwareengineering.stackexchange.com/a/75929)to the singular/plural packages naming)

File: src/main/java/com/example/springlabs/controllers/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  @PutMapping("/**")
+  @ResponseStatus(HttpStatus.OK)
+  public CategoryDto updateCategory(HttpServletRequest request,
+      @RequestBody CategoryDto categoryDto) {
+    List<String> categoriesNames = new ArrayList<>(getCategoriesNames(request));
+    String id = categoriesNames.remove(categoriesNames.size() - 1);
+    return categoryDtoMapper.createDto(
+        categoryService.updateCategory(categoriesNames, Long.parseLong(id),
+            categoryDtoMapper.createCategoryFromDto(categoryDto)));
+  }
```
[the remark this answers is not in the document]
Comment (knurff) [reply src/main/java/com/example/springlabs/controllers/CategoryController.java]: Renamed

File: src/main/java/com/example/springlabs/controllers/CategoryController.java (hunk 2 above)
Comment (knurff) [reply src/main/java/com/example/springlabs/controllers/CategoryController.java]: Renamed

**Review by xd6d — APPROVED** [review]

**Pull request merged**